### DISCUSSION TITLE
Getting dial tcp 127.0.0.1:63509 connect: connection refused

### DISCUSSION THREAD

**Discussion by om3171991** [fluxcd/flux2#1613] [Q&A]
Hey,

I am trying to deploy an application from docker-desktop cluster (all flux v2 components are installed here) to a locally hosted kind cluster by giving kubeconfig in helmRelease (following remote cluster/ cluster-API section from https://fluxcd.io/docs/components/helm/helmreleases/#remote-clusters--cluster-api), but the reconciliation of helm release is failing with below error message :

2021-07-11T14:19:41.231Z info HelmRelease/prod-helm-demo-app.flux-system - Dispatching event: reconciliation failed: failed to get last release revision: query: failed to query with labels: Get &#34;https://127.0.0.1:63509/api/v1/namespaces/prod-helm-demo-app/secrets?labelSelector=name%!D(MISSING)prod-helm-demo-app%!C(MISSING)owner%!D(MISSING)helm&#34;: dial tcp 127.0.0.1:63509: connect: connection refused 
2021-07-11T14:19:41.320Z info HelmRelease/prod-helm-demo-app.flux-system - reconcilation finished in 203.1449ms, next run in 30s 
**2021-07-11T14:19:41.320Z error HelmRelease/prod-helm-demo-app.flux-system - Reconciler error failed to get last release revision: query: failed to query with labels: Get &#34;https://127.0.0.1:63509/api/v1/namespaces/prod-helm-demo-app/secrets?labelSelector=name%3Dprod-helm-demo-app%2Cowner%3Dhelm&#34;: dial tcp 127.0.0.1:63509: connect: connection refused**
2021-07-11T14:19:42.135Z info GitRepository/production.flux-system - Reconciliation finished in 1.3522838s, next run in 30s 
2021-07-11T14:19:42.927Z info Kustomization/stage.flux-system - Kustomization applied in 469.3156ms 
2021-07-11T14:19:42.987Z info Kustomization/stage.flux-system - Reconciliation finished in 1.1285804s, next run in 30s 

When I try the above-highlighted API from a local browser using the service account token as provided in kubeconfig secret - I am able to get the required response.

- Can someone guide me with - what am I missing which is causing this error.
- How can we do deployment on other clusters apart from where the flux is installed to anticipate redundant flux installation. I have already seen CAPI way and currently evaluating the same. if someone knows any other - it will really a help. 

**Update** : 

I have deployed the test AKS cluster and updated the helm release kubeconfig secret and which fixed above error message but still, the application is not deployed in AKS cluster and running reconcile helm release gives another error message : 

2021-07-12T15:56:46.938Z info HelmRelease/prod-helm-demo-app.flux-system - reconciliation finished in 742.2284ms, next run in 30s 
2021-07-12T15:56:46.938Z error HelmRelease/prod-helm-demo-app.flux-system - **Reconciler error failed to get last release revision: query: failed to query with labels: Unauthorized**

The error message seems to be incomplete as to where & which label is missing. Please find below HelmRelease file : 

---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: helm-demo-app
  namespace: flux-system
spec:
  chart:
    spec:
      chart: ./helm/demoapp
      sourceRef:
        kind: GitRepository
        name: production
  install:
    createNamespace: true
    remediation:
      retries: -1
  upgrade:
    remediation:
      retries: -1
  interval: 30s
  timeout: 2m
  releaseName: helm-demo-app
  targetNamespace: helm-demo-app
  storageNamespace: helm-demo-app
  kubeConfig:
    secretRef:
      name: devops-kubeconfig
  values:
    env: helm
    replicaCount: 1

if somebody has any idea what am I doing wrong or any working example - Please help.

cc - @stealthybox @stefanprodan @yebyen @kingdonb

**Comment by kingdonb** [comment]
> from docker-desktop cluster (all flux v2 components are installed here) to a locally hosted kind cluster

The 127.0.0.1 address is a localhost/loopback address. This is a local, unrouted address, which exists in every pod and on your local machine, but they are not the same 127.0.0.1. If you copied the kubeconfig from one locally-hosted cluster to another, this endpoint will definitely not be reachable from the pod network inside of either cluster.

In a normal cluster, the `control-plane-endpoint` is either the public address of a master node, or for HA clusters, the address of a Load Balancer which represents the publicly addressable place where the control plane can be reached by authorized users, (like the remote Flux user with the copied kubeconfig.)

You will not have this same issue with CAPI clusters or other clusters that are designed for production use, or clusters that are hosted on a real (routed) network, since they will not have their control plane endpoints published at 127.0.0.1. The kind cluster does this because it is not intended for production use.

This can be configured in the kind cluster options, this is the quickest way to facilitate your experiment without starting over, you can do so as described here: https://kind.sigs.k8s.io/docs/user/configuration/#api-server

**Reply by om3171991** [reply]
@kingdonb - Thanks and appreciate your help. I tried config as given in kind doc but it didn't worked.

**Reply by yebyen** [reply]
The kind example config linked only shows how to make the implicit config explicit. It still used 127.0.0.1, this doesn't actually change anything.

What you need is one cluster whose control plane address is addressible to the other cluster. Try setting the value to your machine's NAT address of the local private network.

Another good test is: can you take the kubeconfig file to another machine on the local network and use the cluster from there (with kubectl)? If so, then it is more likely the flux instance on your docker desktop cluster will also be able to use it.

**Reply by om3171991** [reply]
@yebyen : Thanks for the reply. I am little new to this Kubernetes and networking ecosystem but what I understand and inference from a reply - I should give it try with IPv4 address present in ifconfig output. Please correct me if wrong and if correct then also it is not working.
